### PR TITLE
feat(medusa): Improve product update performances

### PR DESCRIPTION
**Pull request by adrien2p**
- create 15 products during setup, each product has 50 variants with each a pricing
- in each request, update the amount of each price for each variant of the product being updated

For simple cases where no props on the variants need to be updated or the prices amount did not changed, the update are avoided and therefore also improve by removing unnecessary steps in the flow. 

A few methods have been migrated to bulk to avoid unnecessary queries and reduce the amount of those queries.

Other improvements have been added overall and you can look at the pr itself to get a better understanding.

**NOTE**
https://github.com/medusajs/medusa-staging/pull/14

**Test**
```js
// load testing config
export const scenarios = {
  scenarios: {
    updateProductVariantPrices: {
	  executor: 'ramping-arrival-rate',
	  startTime: '3s',
      startRate: 5,
      timeUnit: '1s', // we start at 5 iterations per second
	  stages: [
        { target: 15, duration: '5s' },
        { target: 15, duration: '5s' },
        { target: 0, duration: '5s' },
      ],
      preAllocatedVUs: 20,
      maxVUs: 50, // if the preAllocatedVUs are not enough, we can initialize more
      exec: "updateProductVariantPrices_"
    }
  }
};
```

The tests have been ran on a single instance on my local machine, the important thing to look at is the factor between 
before/after more than the numbers themselves as it can change depending on the resources.

--

The data used are the same shape for both tests against master and against the pr
- 15 products created
- 50 variants per product
- 1 price per variant
Each request will update one product including all the variants and update the price of each variant, the data looks like this
```js
// data to update in the post body
{
  variants: [
    // 49 other variants
    {
      id: ...,
      prices: [{
        currency_code: "usd",
        amount: 20000,
      }]
    }
  }]
}
```

You can see the results in the images with left (master) and right (the pr)

image (1)
image

RESOLVE CORE-1208

### CODE REVIEW

**Comment by adrien2p** [comment]
@olivermrbl once bulk emit is merged, I ll be able to update that pr with it

File: packages/medusa/src/services/__tests__/product-variant.js [hunk 1]
```diff
@@ -718,33 +766,31 @@ describe("ProductVariantService", () => {
       expect(moneyAmountRepository.create).toHaveBeenCalledWith({
         variant_id: IdMap.getId("ironman"),
         region_id: IdMap.getId("cali"),
+        currency_code: "usd",
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/product-variant.js:769]: self-review: for some reason, on master the currency_code did not appeared even though it is a non nullable column and the the old setRegionPrices was having the currency_code, do I miss something?

File: packages/medusa/src/services/__tests__/product-variant.js [hunk 2]
```diff
@@ -718,33 +766,31 @@ describe("ProductVariantService", () => {
...
-      ).toHaveBeenCalledWith(IdMap.getId("ironman"), {
-        id: IdMap.getId("dkk"),
-        currency_code: "dkk",
-        amount: 750,
-      })
+      expect(moneyAmountRepository.update).toHaveBeenCalledTimes(1)
+      expect(moneyAmountRepository.update).toHaveBeenCalledWith(
+        { id: IdMap.getId("dkk") },
+        {
+          amount: 850,
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/product-variant.js:791]: self-review: Here only the amount changed, this is why the currecny_cde does not appear anymore

File: packages/medusa/src/api/routes/admin/products/update-product.ts [hunk 3]
```diff
@@ -125,31 +133,74 @@ export default async (req, res) => {
...
     }
 
-    const product = await productService
+    const variantRepo = manager.getCustomRepository(productVariantRepo)
+    const productVariants = await productVariantService
       .withTransaction(transactionManager)
-      .retrieve(id, {
-        relations: ["variants"],
-      })
+      .list(
```
Comment (riqwan) [inline packages/medusa/src/api/routes/admin/products/update-product.ts:150]: question: wouldn't this be paginated? Are we safeguading cases where product variants can be > 20?

File: packages/medusa/src/repositories/image.ts [hunk 4]
```diff
@@ -1,8 +1,25 @@
 import { EntityRepository, In, Repository } from "typeorm"
-import { Image } from "../models/image"
+import { Image } from "../models"
+import { QueryDeepPartialEntity } from "typeorm/query-builder/QueryPartialEntity"
 
 @EntityRepository(Image)
 export class ImageRepository extends Repository<Image> {
+  async insertBulk(data: QueryDeepPartialEntity<Image>[]): Promise<Image[]> {
```
Comment (riqwan) [inline packages/medusa/src/repositories/image.ts:7]: `insertBulk` would imply that this is doing a bulk insert, but this looks just like an createMany instead

File: packages/medusa/src/services/product-variant.ts [hunk 5]
```diff
@@ -245,86 +254,197 @@ class ProductVariantService extends TransactionBaseService {
...
+    update: UpdateProductVariantInput
+  ): Promise<ProductVariant>
+
+  async update<
+    TInput extends string | Partial<ProductVariant> | UpdateProductVariantData,
+    TResult = TInput extends UpdateProductVariantData
+      ? ProductVariant[]
+      : ProductVariant
+  >(
+    variantOrVariantIdOrData: TInput,
```
Comment (riqwan) [inline packages/medusa/src/services/product-variant.ts:293]: suggestion: can we split this into 2 or 3 different variables :D the dance we do to identify what this variable is underneath is pretty hard to follow.

Comment (adrien2p) [reply packages/medusa/src/services/product-variant.ts:293]: This is to be compatible, can't do that. Also, the overload allow you to know what the argument can be pretty easily, the implemented method must reflect all overload

File: packages/medusa/src/services/product-variant.ts [hunk 6]
```diff
@@ -245,86 +254,197 @@ class ProductVariantService extends TransactionBaseService {
...
+              }
+            }
+
+            let result = variant
+
+            // No need to update if the nothing on the variant has been changed
+            if (shouldUpdate) {
+              const { id } = variant
+              const rawResult = await variantRepo.update({ id }, toUpdate)
+              result = variantRepo.create({
```
Comment (riqwan) [inline packages/medusa/src/services/product-variant.ts:409]: question: why are we creating after we update?

File: packages/medusa/src/utils/__tests__/has-changes.spec.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import { hasChanges } from "../has-changes";
+
+describe('hasChanges', function () {
```
Comment (olivermrbl) [inline packages/medusa/src/utils/__tests__/has-changes.spec.ts:3]: question: Is this linted correctly?

File: packages/medusa/src/strategies/batch-jobs/product/import.ts [hunk 8]
```diff
@@ -584,12 +581,13 @@ class ProductImportStrategy extends AbstractBatchJobStrategy {
 
         await this.prepareVariantOptions(variantOp, product.id)
 
+        const updateData = transformVariantData(variantOp)
+        delete updateData.product
```
Comment (olivermrbl) [inline packages/medusa/src/strategies/batch-jobs/product/import.ts:585]: question: What led to this change? Just curious, as it seems unrelated to the PR scope

File: packages/medusa/src/services/product-variant.ts [hunk 9]
```diff
@@ -245,86 +254,197 @@ class ProductVariantService extends TransactionBaseService {
...
+              ) as (keyof ProductVariant)[],
+            }
           )
-        } else {
-          variant = variantRes as ProductVariant
+          if (!isDefined(variantRes)) {
+            throw new MedusaError(
+              MedusaError.Types.NOT_FOUND,
+              `Variant with id ${variantOrVariantIdOrData} was not found`
+            )
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant.ts:322]: suggestion(non-blocking): Use `this.retrieve()` with baked in error handling

File: packages/medusa/src/services/product-variant.ts [hunk 10]
```diff
@@ -245,86 +254,197 @@ class ProductVariantService extends TransactionBaseService {
...
-      }
+      const result = await this.updateBatch(data)
+
+      return (Array.isArray(variantOrVariantIdOrData)
+        ? result
+        : result[0]) as unknown as TResult
+    })
+  }
+
+  protected async updateBatch(
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant.ts:346]: nit: Should we rename to `updateBulk` to be consistent with our repositories?

File: packages/medusa/src/api/routes/admin/products/update-product.ts (hunk 3 above)
Comment (riqwan) [reply packages/medusa/src/api/routes/admin/products/update-product.ts:150]: Maybe I'm missing something here. But doesn't this mean that its paginated? How are we skipping pagination here?

```
async list(
    selector: FilterableProductVariantProps,
    config: FindConfig<ProductVariant> & PriceSelectionContext = {
      relations: [],
      skip: 0,
      take: 20,
    }
  )
```

**Review by olivermrbl — COMMENTED** [review]
Couple of initial comments. Haven't gotten to the variant service changes yet :)

File: packages/medusa/src/services/product-variant.ts (hunk 10 above)
Comment (adrien2p) [reply packages/medusa/src/services/product-variant.ts:346]: generally we are using `batch`, I ve only used `bulk` to distinguish the original insert from the custom one and to not call the custom one `insert` in order to not override the original implementation

Comment (olivermrbl) [reply packages/medusa/src/services/product-variant.ts:346]: Got it. Fine to keep it as is :)

File: packages/medusa/src/strategies/batch-jobs/product/import.ts (hunk 8 above)
Comment (adrien2p) [reply packages/medusa/src/strategies/batch-jobs/product/import.ts:585]: It is unrelated, but I encounter this one as it is not expected by the update method, product should not exists, otherwise if a cascade occurs it might update the product where you don't expect it to be

File: packages/medusa/src/utils/__tests__/has-changes.spec.ts (hunk 7 above)
Comment (adrien2p) [reply packages/medusa/src/utils/__tests__/has-changes.spec.ts:3]: the file?

File: packages/medusa/src/services/product-variant.ts (hunk 6 above)
Comment (adrien2p) [reply packages/medusa/src/services/product-variant.ts:409]: it create a new instance of an entity, it is not create in the sense of insert, it is more of a factory

File: packages/medusa/src/api/routes/admin/products/update-product.ts (hunk 3 above)
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/products/update-product.ts:150]: When you pass a config object, the default value does not take place since there is a value that has been passed

File: packages/medusa/src/services/product-variant.ts (hunk 6 above)
Comment (adrien2p) [reply packages/medusa/src/services/product-variant.ts:398]: for reference:

```ts
create(entityClass, plainObjectOrObjects) {
        const metadata = this.connection.getMetadata(entityClass);
        if (!plainObjectOrObjects)
            return metadata.create(this.queryRunner);
        if (Array.isArray(plainObjectOrObjects))
            return plainObjectOrObjects.map((plainEntityLike) => this.create(entityClass, plainEntityLike));
        const mergeIntoEntity = metadata.create(this.queryRunner);
        this.plainObjectToEntityTransformer.transform(mergeIntoEntity, plainObjectOrObjects, metadata, true);
        return mergeIntoEntity;
    }
```

File: packages/medusa/src/utils/__tests__/has-changes.spec.ts (hunk 7 above)
Comment (adrien2p) [reply packages/medusa/src/utils/__tests__/has-changes.spec.ts]: fixed

File: packages/medusa/src/services/product-variant.ts [hunk 11]
```diff
@@ -245,86 +254,186 @@ class ProductVariantService extends TransactionBaseService {
...
+              for (const [key, value] of Object.entries(rest)) {
+                if (variant[key] !== value) {
+                  toUpdate[key] = value
+                }
+              }
+            }
+
+            let result = variant
+
+            // No need to update if the nothing on the variant has been changed
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant.ts:394]:
```suggestion
            // No need to update if nothing on the variant has changed
```

File: packages/medusa/src/services/product-variant.ts [hunk 12]
```diff
@@ -245,86 +254,186 @@ class ProductVariantService extends TransactionBaseService {
...
+        .filter(([, , shouldEmitUpdateEvent]) => shouldEmitUpdateEvent)
+        .map(([result, updatedData]) => {
+          return {
+            eventName: ProductVariantService.Events.UPDATED,
+            data: {
+              id: result.id,
+              product_id: result.product_id,
+              fields: Object.keys(updatedData),
+            },
+          }
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant.ts:418]: praise: I like this!

**Review by olivermrbl — APPROVED** [review]
LGTM - this is a great piece of work @adrien2p!

nitpicking: Should we create types for some of these service method payloads? Think it is slightly messy with inline typings

**Comment by olivermrbl** [comment]
@riqwan Will you give this a final review too? 🙏

**Comment by olivermrbl** [comment]
@adrien2p Please add a changeset :)

**Review by riqwan — APPROVED** [review]
This must've been a real test, looks great! 🎉

**Comment by adrien2p** [comment]
> LGTM - this is a great piece of work @adrien2p!
> 
> nitpicking: Should we create types for some of these service method payloads? Think it is slightly messy with inline typings


I ve includes the types @olivermrbl
